lib/utils/compose_ts.ts

```diff
@@ -1493,10 +1494,15 @@ function pullProgressAdapter(outStream: Duplex) {
 		error: Error;
 		errorDetail: Error;
 	}) {
-		if (status != null) {
+		id ||= '';
+		status ||= '';
+		const isTotal = id && id.toLowerCase() === 'total';
+		if (status) {
 			status = status.replace(/^Status: /, '');
+		} else if (isTotal && typeof percentage === 'number') {
+			status = `Pull progress: ${percentage}%`;
 		}
-		if (id != null) {
+		if (id && status && !isTotal) {
```

Comment:
It took me a while to track down where these values were coming from but it looks like it's from this build log task?
https://github.com/balena-io/balena-cli/blob/4911db640f4979c8561ba1a4b8d103bcf7db4ae2/lib/utils/device/deploy.ts#L485-L489

Is it parsing raw logs from a docker pull command?

Reply:
`utils/device/deploy.ts` is a different code path, used by `balena push <ip-addr>`.
The changes in this commit affect `balena build`.

The `id` and `percentage` fields are assigned in the following function:
https://github.com/balena-io-modules/docker-progress/blob/v5.0.0/lib/index.ts#L332-L348

The source is indeed logs from a `docker pull` command (API call through `dockerode`), but not raw, as it is a second layer of parsing after the `docker-progress` module has already done some parsing/transforming of its own:

* `task.progressHook = pullProgressAdapter(captureStream);`
at https://github.com/balena-io/balena-cli/blob/v12.44.12/lib/utils/compose_ts.ts#L496

* `const progressHook = task.progressHook;`  
`return dockerProgress.pull(imageName, progressHook, opts);`  
at https://github.com/balena-io-modules/balena-multibuild/blob/v4.11.0/lib/external.ts#L68

* `onProgress = reporter.pullProgress(onProgress);`
at https://github.com/balena-io-modules/docker-progress/blob/v5.0.0/lib/index.ts#L428

By the way, `Total: undefined` is nothing introduced (nor fixed) by `docker-progress` v5, this bug was around for a long time, maybe longer than a year.

Before this commit (and for a long time, maybe longer than a year):
```
$ balena build ~/balena/other/simple_multicontainer -h 192.168.1.46 -p 2375 -a test-nuc
[Info]    Building for amd64/intel-nuc
[Build]   Building services...
[Build]   foo Preparing...
[Build]   bar Preparing...
[Build]   zip Preparing...
...
[Build]   foo Total: undefined
[Build]   foo Total: undefined
[Build]   foo Total: undefined
...
```
After this commit:
```
...
[Build]   foo Pull progress: 54%
[Build]   foo Pull progress: 56%
[Build]   foo Pull progress: 59%
...
```